mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java

```diff
@@ -1512,4 +1549,101 @@ void searchShouldRetrieveMailByAttachmentFileName() throws Exception {
+            .collectList().block();
+
+        assertThat(actual).isEqualTo(ImmutableList.of(message1.getMessageId(), message2.getMessageId(), message3.getMessageId()));
+    }
+
+    @Test
+    void givenAMailInAThreadThenGetThreadShouldReturnAListWithOnlyOneMessageIdInThatThread() throws MailboxException {
+        MailboxMessage message1 = createMessage(quanMailbox, ThreadId.fromBaseMessageId(newBasedMessageId));
+
+        appendMessageThenDispatchAddedEvent(quanMailbox, message1);
+
+        awaitMessageCount(ImmutableList.of(), SearchQuery.matchAll(), 14);
+
+        SearchQuery searchQuery = SearchQuery.of(SearchQuery.threadId(ThreadId.fromBaseMessageId(newBasedMessageId)));
+        List<MessageId> actual = messageSearchIndex.search(quanSession, ImmutableList.of(quanMailbox.getMailboxId()), searchQuery, LIMIT)
+            .collectList().block();
+
+        assertThat(actual).containsOnly(message1.getMessageId());
+    }
+
+    @Test
+    void givenTwoDistinctThreadsThenGetThreadShouldNotReturnUnrelatedMails() throws MailboxException {
+        // given message1 and message2 in thread1, message3 in thread2
+        ThreadId threadId1 = ThreadId.fromBaseMessageId(newBasedMessageId);
+        ThreadId threadId2 = ThreadId.fromBaseMessageId(otherBasedMessageId);
+        MailboxMessage message1 = createMessage(quanMailbox, threadId1);
+        MailboxMessage message2 = createMessage(quanMailbox, threadId1);
+        MailboxMessage message3 = createMessage(quanMailbox, threadId2);
+
+        appendMessageThenDispatchAddedEvent(quanMailbox, message1);
+        appendMessageThenDispatchAddedEvent(quanMailbox, message2);
+        appendMessageThenDispatchAddedEvent(quanMailbox, message3);
+
+        awaitMessageCount(ImmutableList.of(), SearchQuery.matchAll(), 16);
+
+        // then get thread2 should not return unrelated message1 and message2
+        SearchQuery searchQuery = SearchQuery.of(SearchQuery.threadId(threadId2));
+        List<MessageId> actual = messageSearchIndex.search(quanSession, ImmutableList.of(quanMailbox.getMailboxId()), searchQuery, LIMIT)
+            .collectList().block();
+
+        assertThat(actual).doesNotContain(message1.getMessageId(), message2.getMessageId());
+    }
+
+    @Test
+    void givenNonThreadThenGetThreadShouldReturnEmptyListMessageId() throws MailboxException {
+        // given non messages in thread1
+        ThreadId threadId1 = ThreadId.fromBaseMessageId(newBasedMessageId);
+
+        // then get thread1 should return empty list messageId
+        SearchQuery searchQuery = SearchQuery.of(SearchQuery.threadId(threadId1));
+        List<MessageId> actual = messageSearchIndex.search(quanSession, ImmutableList.of(quanMailbox.getMailboxId()), searchQuery, LIMIT)
+            .collectList().block();
+
+        assertThat(actual).isEmpty();
+    }
+
+    private void appendMessageThenDispatchAddedEvent(Mailbox mailbox, MailboxMessage mailboxMessage) throws MailboxException {
```

Comment:
GIVEN that this class can rely on the `scanning thread id guessing algorithm` (as InMemoryIntegrationResources should anyway),

Can't we instead append messages to the main storeMailboxManager that happen to share (or not share) the `In-Reply-To` ?

This would remove all of the mapper dark magic, make us more feature oriented and also make this testing code more bullet proof.

My 2 cents.

Reply:
I know that. But how we solve that is an issue now.
Once we satisfy with the scanning threadId guessing algorithm then we can apply that to CassandraModule, JPA module, InmemoryModule by moving ThreadIdGuessingAlgorithm APIs to MailboxManager. That's will solve the constructor loop.

Is this scanning with ElasticSearch good enough for production? Until then we maybe can move on.

Reply:
If it's not yet. I guess we can come back and remove these message mapper dark magic later?

Reply:
I would like that we have a shot at it now if you do not mind ;-)

Reply:
Fair